File: go/mechanisms/evm/erc4337/alchemy.go
package erc4337

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"math/big"
	"net/http"
	"time"

	"github.com/ethereum/go-ethereum/common"
)

// AlchemyConfig contains configuration for Alchemy bundler client.
type AlchemyConfig struct {
	// APIKey is the Alchemy API key
	APIKey string `json:"apiKey"`
	// ChainID is the chain ID
	ChainID int64 `json:"chainId"`
	// BundlerURL is an optional custom bundler URL
	BundlerURL string `json:"bundlerUrl,omitempty"`
	// EntryPoint is the EntryPoint contract address (optional, defaults to v0.7)
	EntryPoint common.Address `json:"entryPoint,omitempty"`
	// PolicyID is the Alchemy gas manager policy ID for sponsorship
	PolicyID string `json:"policyId,omitempty"`
}

// AssetChange represents an asset change from simulation.
type AssetChange struct {
	// AssetType: native, erc20, erc721, erc1155
	AssetType string `json:"assetType"`
	// ChangeType: transfer_in, transfer_out
	ChangeType string `json:"changeType"`
	// From account
	From common.Address `json:"from"`
	// To account
	To common.Address `json:"to"`
	// Amount (for native/erc20) or empty
	Amount *big.Int `json:"amount,omitempty"`
	// TokenID (for erc721)
	TokenID *big.Int `json:"tokenId,omitempty"`
	// ContractAddress (for tokens)
	ContractAddress common.Address `json:"contractAddress,omitempty"`
	// Symbol
	Symbol string `json:"symbol,omitempty"`
	// Name
	Name string `json:"name,omitempty"`
	// Decimals
	Decimals int `json:"decimals,omitempty"`
}

// SimulationResult contains the result of asset simulation.
type SimulationResult struct {
	// Success indicates if simulation succeeded
	Success bool `json:"success"`
	// Error message if failed
	Error string `json:"error,omitempty"`
	// Changes from the operation
	Changes []AssetChange `json:"changes"`
}

// GasAndPaymasterResult contains combined gas + paymaster data.
type GasAndPaymasterResult struct {
	// GasEstimate from the bundler
	GasEstimate *GasEstimate `json:"gasEstimate"`
	// PaymasterData if sponsorship is enabled
	PaymasterData *PaymasterData `json:"paymasterData,omitempty"`
	// MaxFeePerGas recommended
	MaxFeePerGas *big.Int `json:"maxFeePerGas"`
	// MaxPriorityFeePerGas recommended
	MaxPriorityFeePerGas *big.Int `json:"maxPriorityFeePerGas"`
}

// AlchemyBundlerClient is an Alchemy-specific bundler client.
type AlchemyBundlerClient struct {
	*GenericBundlerClient
	apiKey     string
	alchemyURL string
	policyID   string
}

// NewAlchemyBundlerClient creates a new Alchemy bundler client.
func NewAlchemyBundlerClient(config AlchemyConfig) *AlchemyBundlerClient {
	bundlerURL := config.BundlerURL
	if bundlerURL == "" {
		network, _ := GetAlchemyNetwork(config.ChainID)
		bundlerURL = fmt.Sprintf("https://%s.g.alchemy.com/v2/%s", network, config.APIKey)
	}

	entryPoint := config.EntryPoint
	if entryPoint == (common.Address{}) {
		entryPoint = common.HexToAddress(EntryPointV07Address)
	}

	genericClient := NewBundlerClient(BundlerConfig{
		BundlerURL: bundlerURL,
		EntryPoint: entryPoint,
		ChainID:    config.ChainID,
	})

	return &AlchemyBundlerClient{
		GenericBundlerClient: genericClient,
		apiKey:               config.APIKey,
		alchemyURL:           bundlerURL,
		policyID:             config.PolicyID,
	}
}

// RequestGasAndPaymasterAndData requests gas estimates and paymaster data in one call.
func (c *AlchemyBundlerClient) RequestGasAndPaymasterAndData(
	userOp *UserOperation,
	overrides *GasOverrides,
) (*GasAndPaymasterResult, error) {
	if c.policyID == "" {
		return nil, fmt.Errorf("Alchemy policy ID required for gas sponsorship")
	}

	packed := c.packPartialUserOp(userOp)

	request := map[string]interface{}{
		"policyId":      c.policyID,
		"entryPoint":    c.entryPoint.Hex(),
		"userOperation": packed,
		"dummySignature": getDummySignature(),
	}

	if overrides != nil {
		overridesMap := make(map[string]string)
		if overrides.MaxFeePerGas != nil {
			overridesMap["maxFeePerGas"] = bigIntToHex(overrides.MaxFeePerGas)
		}
		if overrides.MaxPriorityFeePerGas != nil {
			overridesMap["maxPriorityFeePerGas"] = bigIntToHex(overrides.MaxPriorityFeePerGas)
		}
		if overrides.CallGasLimit != nil {
			overridesMap["callGasLimit"] = bigIntToHex(overrides.CallGasLimit)
		}
		if overrides.VerificationGasLimit != nil {
			overridesMap["verificationGasLimit"] = bigIntToHex(overrides.VerificationGasLimit)
		}
		if overrides.PreVerificationGas != nil {
			overridesMap["preVerificationGas"] = bigIntToHex(overrides.PreVerificationGas)
		}
		if len(overridesMap) > 0 {
			request["overrides"] = overridesMap
		}
	}

	var result struct {
		PaymasterAndData     string `json:"paymasterAndData"`
		CallGasLimit         string `json:"callGasLimit"`
		VerificationGasLimit string `json:"verificationGasLimit"`
		PreVerificationGas   string `json:"preVerificationGas"`
		MaxFeePerGas         string `json:"maxFeePerGas"`
		MaxPriorityFeePerGas string `json:"maxPriorityFeePerGas"`
	}

	err := c.alchemyRPCCall("alchemy_requestGasAndPaymasterAndData", []interface{}{request}, &result)
	if err != nil {
		return nil, err
	}

	gasResult := &GasAndPaymasterResult{
		GasEstimate: &GasEstimate{
			VerificationGasLimit: hexToBigInt(result.VerificationGasLimit),
			CallGasLimit:         hexToBigInt(result.CallGasLimit),
			PreVerificationGas:   hexToBigInt(result.PreVerificationGas),
		},
		MaxFeePerGas:         hexToBigInt(result.MaxFeePerGas),
		MaxPriorityFeePerGas: hexToBigInt(result.MaxPriorityFeePerGas),
	}

	// Parse paymasterAndData into PaymasterData
	if result.PaymasterAndData != "" && result.PaymasterAndData != "0x" && len(result.PaymasterAndData) >= 106 {
		paymasterData := &PaymasterData{
			Paymaster: common.HexToAddress(result.PaymasterAndData[0:42]),
		}
		// Parse verification gas limit (bytes 20-36)
		if len(result.PaymasterAndData) >= 74 {
			paymasterData.PaymasterVerificationGasLimit = hexToBigInt(result.PaymasterAndData[42:74])
		}
		// Parse post-op gas limit (bytes 36-52)
		if len(result.PaymasterAndData) >= 106 {
			paymasterData.PaymasterPostOpGasLimit = hexToBigInt(result.PaymasterAndData[74:106])
		}
		// Parse data (bytes 52+)
		if len(result.PaymasterAndData) > 106 {
			paymasterData.PaymasterData = hexToBytes(result.PaymasterAndData[106:])
		}
		gasResult.PaymasterData = paymasterData
		gasResult.GasEstimate.PaymasterVerificationGasLimit = paymasterData.PaymasterVerificationGasLimit
		gasResult.GasEstimate.PaymasterPostOpGasLimit = paymasterData.PaymasterPostOpGasLimit
	}

	return gasResult, nil
}

// SimulateUserOperationAssetChanges simulates asset changes from a UserOperation.
func (c *AlchemyBundlerClient) SimulateUserOperationAssetChanges(userOp *UserOperation) (*SimulationResult, error) {
	packed := c.packUserOp(userOp)

	request := map[string]interface{}{
		"entryPoint":    c.entryPoint.Hex(),
		"userOperation": packed,
	}

	var result struct {
		Changes []struct {
			AssetType       string `json:"assetType"`
			ChangeType      string `json:"changeType"`
			From            string `json:"from"`
			To              string `json:"to"`
			Amount          string `json:"amount,omitempty"`
			TokenID         string `json:"tokenId,omitempty"`
			ContractAddress string `json:"contractAddress,omitempty"`
			Symbol          string `json:"symbol,omitempty"`
			Name            string `json:"name,omitempty"`
			Decimals        int    `json:"decimals,omitempty"`
		} `json:"changes"`
	}

	err := c.alchemyRPCCall("alchemy_simulateUserOperationAssetChanges", []interface{}{request}, &result)
	if err != nil {
		return &SimulationResult{
			Success: false,
			Error:   err.Error(),
			Changes: []AssetChange{},
		}, nil
	}

	changes := make([]AssetChange, len(result.Changes))
	for i, change := range result.Changes {
		changes[i] = AssetChange{
			AssetType:       change.AssetType,
			ChangeType:      change.ChangeType,
			From:            common.HexToAddress(change.From),
			To:              common.HexToAddress(change.To),
			Symbol:          change.Symbol,
			Name:            change.Name,
			Decimals:        change.Decimals,
		}
		if change.Amount != "" {
			changes[i].Amount = hexToBigInt(change.Amount)
		}
		if change.TokenID != "" {
			changes[i].TokenID = hexToBigInt(change.TokenID)
		}
		if change.ContractAddress != "" {
			changes[i].ContractAddress = common.HexToAddress(change.ContractAddress)
		}
	}

	return &SimulationResult{
		Success: true,
		Changes: changes,
	}, nil
}

// GetFeeHistory gets fee history for gas estimation.
func (c *AlchemyBundlerClient) GetFeeHistory() (*FeeHistory, error) {
	var result struct {
		BaseFeePerGas []string   `json:"baseFeePerGas"`
		Reward        [][]string `json:"reward"`
	}

	err := c.alchemyRPCCall("eth_feeHistory", []interface{}{"0x5", "latest", []int{25, 50, 75}}, &result)
	if err != nil {
		return nil, err
	}

	if len(result.BaseFeePerGas) == 0 {
		return nil, fmt.Errorf("empty fee history response")
	}

	latestBaseFee := hexToBigInt(result.BaseFeePerGas[len(result.BaseFeePerGas)-1])

	var medianPriorityFee *big.Int
	if len(result.Reward) > 0 {
		midIdx := len(result.Reward) / 2
		if len(result.Reward[midIdx]) > 1 {
			medianPriorityFee = hexToBigInt(result.Reward[midIdx][1])
		} else if len(result.Reward[midIdx]) > 0 {
			medianPriorityFee = hexToBigInt(result.Reward[midIdx][0])
		}
	}
	if medianPriorityFee == nil {
		medianPriorityFee = big.NewInt(1000000000) // 1 gwei default
	}

	// maxFeePerGas = baseFee * 2 + priorityFee
	maxFeePerGas := new(big.Int).Mul(latestBaseFee, big.NewInt(2))
	maxFeePerGas.Add(maxFeePerGas, medianPriorityFee)

	return &FeeHistory{
		BaseFeePerGas:        latestBaseFee,
		MaxPriorityFeePerGas: medianPriorityFee,
		MaxFeePerGas:         maxFeePerGas,
	}, nil
}

// FeeHistory contains fee history data.
type FeeHistory struct {
	BaseFeePerGas        *big.Int `json:"baseFeePerGas"`
	MaxPriorityFeePerGas *big.Int `json:"maxPriorityFeePerGas"`
	MaxFeePerGas         *big.Int `json:"maxFeePerGas"`
}

// GasOverrides contains optional gas overrides for estimation.
type GasOverrides struct {
	MaxFeePerGas         *big.Int
	MaxPriorityFeePerGas *big.Int
	CallGasLimit         *big.Int
	VerificationGasLimit *big.Int
	PreVerificationGas   *big.Int
}

// packPartialUserOp packs a partial UserOperation for estimation.
func (c *AlchemyBundlerClient) packPartialUserOp(userOp *UserOperation) map[string]interface{} {
	result := map[string]interface{}{
		"sender":   userOp.Sender.Hex(),
		"callData": bytesToHex(userOp.CallData),
	}

	if userOp.Nonce != nil {
		result["nonce"] = bigIntToHex(userOp.Nonce)
	} else {
		result["nonce"] = "0x0"
	}

	if len(userOp.InitCode) > 0 {
		result["initCode"] = bytesToHex(userOp.InitCode)
	} else {
		result["initCode"] = "0x"
	}

	if userOp.VerificationGasLimit != nil {
		result["verificationGasLimit"] = bigIntToHex(userOp.VerificationGasLimit)
	}
	if userOp.CallGasLimit != nil {
		result["callGasLimit"] = bigIntToHex(userOp.CallGasLimit)
	}
	if userOp.PreVerificationGas != nil {
		result["preVerificationGas"] = bigIntToHex(userOp.PreVerificationGas)
	}
	if userOp.MaxFeePerGas != nil {
		result["maxFeePerGas"] = bigIntToHex(userOp.MaxFeePerGas)
	}
	if userOp.MaxPriorityFeePerGas != nil {
		result["maxPriorityFeePerGas"] = bigIntToHex(userOp.MaxPriorityFeePerGas)
	}

	if len(userOp.PaymasterAndData) > 0 {
		result["paymasterAndData"] = bytesToHex(userOp.PaymasterAndData)
	} else {
		result["paymasterAndData"] = "0x"
	}

	if len(userOp.Signature) > 0 {
		result["signature"] = bytesToHex(userOp.Signature)
	} else {
		result["signature"] = getDummySignature()
	}

	return result
}

// alchemyRPCCall makes an Alchemy-specific RPC call.
func (c *AlchemyBundlerClient) alchemyRPCCall(method string, params []interface{}, result interface{}) error {
	c.requestID++

	request := map[string]interface{}{
		"jsonrpc": "2.0",
		"id":      c.requestID,
		"method":  method,
		"params":  params,
	}

	body, err := json.Marshal(request)
	if err != nil {
		return fmt.Errorf("failed to marshal request: %w", err)
	}

	httpClient := &http.Client{Timeout: 30 * time.Second}
	resp, err := httpClient.Post(c.alchemyURL, "application/json", bytes.NewReader(body))
	if err != nil {
		return fmt.Errorf("HTTP request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("HTTP error %d: %s", resp.StatusCode, string(body))
	}

	var response struct {
		Result json.RawMessage `json:"result"`
		Error  *struct {
			Code    int             `json:"code"`
			Message string          `json:"message"`
			Data    json.RawMessage `json:"data,omitempty"`
		} `json:"error"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
		return fmt.Errorf("failed to decode response: %w", err)
	}

	if response.Error != nil {
		return fmt.Errorf("RPC error %d: %s", response.Error.Code, response.Error.Message)
	}

	if result != nil && len(response.Result) > 0 {
		if err := json.Unmarshal(response.Result, result); err != nil {
			return fmt.Errorf("failed to unmarshal result: %w", err)
		}
	}

	return nil
}

// getDummySignature returns a dummy signature for gas estimation.
func getDummySignature() string {
	return "0xfffffffffffffffffffffffffffffff0000000000000000000000000000000007aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa1c"
}

File: typescript/packages/mechanisms/evm/src/erc4337/bundlers/alchemy.ts
/**
 * Alchemy Bundler Client
 *
 * Extended bundler client with Alchemy-specific methods:
 * - alchemy_requestGasAndPaymasterAndData: Combined gas + paymaster estimation
 * - alchemy_simulateUserOperationAssetChanges: Simulate asset changes
 *
 * @see https://docs.alchemy.com/reference/bundler-api-endpoints
 */

import type { Address, Hex } from "viem";
import {
  BundlerClient,
  BundlerError,
} from "../bundler.js";
import type { BundlerConfig } from "../types.js";
import type {
  UserOperation,
  UserOperationResult,
  GasEstimate,
  PaymasterData,
} from "../types.js";
import { ENTRYPOINT_V07_ADDRESS, packAccountGasLimits, packGasFees } from "../constants.js";

/**
 * Alchemy policy ID for gas sponsorship
 */
export interface AlchemyPolicyConfig {
  /** Policy ID from Alchemy dashboard */
  policyId: string;
}

/**
 * Asset change from simulation
 */
export interface AssetChange {
  /** Asset type: native, erc20, erc721, erc1155 */
  assetType: "native" | "erc20" | "erc721" | "erc1155";
  /** Direction: from or to the account */
  changeType: "transfer_in" | "transfer_out";
  /** Account affected */
  from: Address;
  to: Address;
  /** Amount (for native/erc20) or tokenId (for erc721) */
  amount?: string;
  tokenId?: string;
  /** Contract address (for tokens) */
  contractAddress?: Address;
  /** Token symbol/name */
  symbol?: string;
  name?: string;
  decimals?: number;
}

/**
 * Simulation result
 */
export interface SimulationResult {
  /** Whether simulation succeeded */
  success: boolean;
  /** Error message if failed */
  error?: string;
  /** Asset changes from the operation */
  changes: AssetChange[];
}

/**
 * Alchemy bundler configuration
 */
export interface AlchemyConfig {
  /** Bundler RPC URL (optional, auto-generated if not provided) */
  bundlerUrl?: string;
  /** Chain ID */
  chainId: number;
  /** Alchemy API key */
  apiKey: string;
  /** Optional policy configuration for gas sponsorship */
  policy?: AlchemyPolicyConfig;
}

/**
 * Alchemy bundler client with extended methods
 */
export class AlchemyBundlerClient extends BundlerClient {
  private readonly apiKey: string;
  private readonly alchemyUrl: string;
  private readonly policy?: AlchemyPolicyConfig;

  constructor(config: AlchemyConfig) {
    // Construct Alchemy bundler URL
    const network = getAlchemyNetwork(config.chainId);
    const bundlerUrl = config.bundlerUrl && config.bundlerUrl.includes("alchemy")
      ? config.bundlerUrl
      : `https://${network}.g.alchemy.com/v2/${config.apiKey}`;

    super({
      ...config,
      bundlerUrl,
    });

    this.apiKey = config.apiKey;
    this.alchemyUrl = bundlerUrl;
    this.policy = config.policy;
  }

  /**
   * Request gas estimates and paymaster data in a single call
   * This is more efficient than making separate calls
   */
  async requestGasAndPaymasterAndData(
    userOp: Partial<UserOperation> & {
      sender: Address;
      callData: Hex;
    },
    overrides?: {
      maxFeePerGas?: Hex;
      maxPriorityFeePerGas?: Hex;
      callGasLimit?: Hex;
      verificationGasLimit?: Hex;
      preVerificationGas?: Hex;
    },
  ): Promise<{
    gasEstimate: GasEstimate;
    paymasterData?: PaymasterData;
    maxFeePerGas: bigint;
    maxPriorityFeePerGas: bigint;
  }> {
    if (!this.policy) {
      throw new BundlerError("Alchemy policy required for gas sponsorship");
    }

    const packed = this.packPartialUserOp(userOp);

    const result = await this.alchemyRpcCall<{
      paymasterAndData: Hex;
      callGasLimit: Hex;
      verificationGasLimit: Hex;
      preVerificationGas: Hex;
      maxFeePerGas: Hex;
      maxPriorityFeePerGas: Hex;
    }>("alchemy_requestGasAndPaymasterAndData", [
      {
        policyId: this.policy.policyId,
        entryPoint: ENTRYPOINT_V07_ADDRESS,
        userOperation: packed,
        dummySignature: userOp.signature ?? getDummySignature(),
        overrides,
      },
    ]);

    // Parse paymasterAndData into PaymasterData
    let paymasterData: PaymasterData | undefined;
    if (result.paymasterAndData && result.paymasterAndData !== "0x") {
      paymasterData = {
        paymaster: `0x${result.paymasterAndData.slice(2, 42)}` as Address,
        paymasterVerificationGasLimit: BigInt(`0x${result.paymasterAndData.slice(42, 74)}`),
        paymasterPostOpGasLimit: BigInt(`0x${result.paymasterAndData.slice(74, 106)}`),
        paymasterData: `0x${result.paymasterAndData.slice(106)}` as Hex,
      };
    }

    return {
      gasEstimate: {
        verificationGasLimit: BigInt(result.verificationGasLimit),
        callGasLimit: BigInt(result.callGasLimit),
        preVerificationGas: BigInt(result.preVerificationGas),
        paymasterVerificationGasLimit: paymasterData?.paymasterVerificationGasLimit,
        paymasterPostOpGasLimit: paymasterData?.paymasterPostOpGasLimit,
      },
      paymasterData,
      maxFeePerGas: BigInt(result.maxFeePerGas),
      maxPriorityFeePerGas: BigInt(result.maxPriorityFeePerGas),
    };
  }

  /**
   * Simulate asset changes from a UserOperation
   * Useful for previewing what will happen before submitting
   */
  async simulateUserOperationAssetChanges(
    userOp: UserOperation,
  ): Promise<SimulationResult> {
    const packed = this.packUserOpForRpc(userOp);

    try {
      const result = await this.alchemyRpcCall<{
        changes: AssetChange[];
      }>("alchemy_simulateUserOperationAssetChanges", [
        {
          entryPoint: ENTRYPOINT_V07_ADDRESS,
          userOperation: packed,
        },
      ]);

      return {
        success: true,
        changes: result.changes,
      };
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : "Simulation failed",
        changes: [],
      };
    }
  }

  /**
   * Get fee history for gas estimation
   */
  async getFeeHistory(): Promise<{
    baseFeePerGas: bigint;
    maxPriorityFeePerGas: bigint;
    maxFeePerGas: bigint;
  }> {
    // Use standard Alchemy fee history
    const result = await this.alchemyRpcCall<{
      baseFeePerGas: Hex[];
      reward: Hex[][];
    }>("eth_feeHistory", ["0x5", "latest", [25, 50, 75]]);

    const latestBaseFee = BigInt(result.baseFeePerGas[result.baseFeePerGas.length - 1]);
    const medianPriorityFee = BigInt(result.reward[Math.floor(result.reward.length / 2)][1]);

    return {
      baseFeePerGas: latestBaseFee,
      maxPriorityFeePerGas: medianPriorityFee,
      maxFeePerGas: latestBaseFee * 2n + medianPriorityFee,
    };
  }

  /**
   * Send UserOperation with Alchemy optimizations
   */
  async sendUserOperationWithAlchemy(
    userOp: UserOperation,
  ): Promise<UserOperationResult> {
    const packed = this.packUserOpForRpc(userOp);

    const userOpHash = await this.alchemyRpcCall<Hex>(
      "eth_sendUserOperation",
      [packed, ENTRYPOINT_V07_ADDRESS],
    );

    return {
      userOpHash,
      wait: () => this.waitForReceipt(userOpHash),
    };
  }

  /**
   * Pack partial UserOperation for estimation
   */
  private packPartialUserOp(
    userOp: Partial<UserOperation> & { sender: Address; callData: Hex },
  ): Record<string, unknown> {
    return {
      sender: userOp.sender,
      nonce: userOp.nonce !== undefined ? this.bigintToHex(userOp.nonce) : "0x0",
      initCode: userOp.initCode ?? "0x",
      callData: userOp.callData,
      callGasLimit: userOp.callGasLimit !== undefined
        ? this.bigintToHex(userOp.callGasLimit)
        : undefined,
      verificationGasLimit: userOp.verificationGasLimit !== undefined
        ? this.bigintToHex(userOp.verificationGasLimit)
        : undefined,
      preVerificationGas: userOp.preVerificationGas !== undefined
        ? this.bigintToHex(userOp.preVerificationGas)
        : undefined,
      maxFeePerGas: userOp.maxFeePerGas !== undefined
        ? this.bigintToHex(userOp.maxFeePerGas)
        : undefined,
      maxPriorityFeePerGas: userOp.maxPriorityFeePerGas !== undefined
        ? this.bigintToHex(userOp.maxPriorityFeePerGas)
        : undefined,
      paymasterAndData: userOp.paymasterAndData ?? "0x",
      signature: userOp.signature ?? getDummySignature(),
    };
  }

  /**
   * Pack UserOperation for RPC
   */
  private packUserOpForRpc(userOp: UserOperation): Record<string, unknown> {
    return {
      sender: userOp.sender,
      nonce: this.bigintToHex(userOp.nonce),
      initCode: userOp.initCode,
      callData: userOp.callData,
      accountGasLimits: packAccountGasLimits(
        userOp.verificationGasLimit,
        userOp.callGasLimit,
      ),
      preVerificationGas: this.bigintToHex(userOp.preVerificationGas),
      gasFees: packGasFees(userOp.maxPriorityFeePerGas, userOp.maxFeePerGas),
      paymasterAndData: userOp.paymasterAndData,
      signature: userOp.signature,
    };
  }

  /**
   * Convert bigint to hex
   */
  private bigintToHex(value: bigint): Hex {
    return `0x${value.toString(16)}` as Hex;
  }

  /**
   * Make Alchemy RPC call
   */
  private async alchemyRpcCall<T>(method: string, params: unknown[]): Promise<T> {
    const response = await fetch(this.alchemyUrl, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        jsonrpc: "2.0",
        id: Date.now(),
        method,
        params,
      }),
    });

    if (!response.ok) {
      throw new BundlerError(
        `HTTP error: ${response.status} ${response.statusText}`,
      );
    }

    const json = await response.json() as {
      result?: T;
      error?: { code: number; message: string; data?: unknown };
    };

    if (json.error) {
      throw new BundlerError(json.error.message, json.error.code, json.error.data);
    }

    return json.result as T;
  }
}

/**
 * Get Alchemy network name from chain ID
 */
function getAlchemyNetwork(chainId: number): string {
  const networks: Record<number, string> = {
    1: "eth-mainnet",
    11155111: "eth-sepolia",
    137: "polygon-mainnet",
    80001: "polygon-mumbai",
    10: "opt-mainnet",
    420: "opt-goerli",
    42161: "arb-mainnet",
    421613: "arb-goerli",
    8453: "base-mainnet",
    84532: "base-sepolia",
  };

  const network = networks[chainId];
  if (!network) {
    throw new BundlerError(`Unsupported chain ID for Alchemy: ${chainId}`);
  }

  return network;
}

/**
 * Get dummy signature for estimation
 */
function getDummySignature(): Hex {
  return "0xfffffffffffffffffffffffffffffff0000000000000000000000000000000007aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa1c" as Hex;
}

/**
 * Create an Alchemy bundler client
 */
export function createAlchemyBundlerClient(config: AlchemyConfig): AlchemyBundlerClient {
  return new AlchemyBundlerClient(config);
}
